### PR TITLE
Implementing Undersampling, Oversampling, and SMOTE Techniques in Deep Neural Networks

### LINKED ISSUE #7272: [Machine Learning] Implementing undersampling, oversampling and SMOTE techniques to handle imbalanced data in deep neural networks
#### Proposed title of article
[Machine Learning] Implementing undersampling, oversampling and SMOTE techniques to handle imbalanced data in deep neural networks

#### Proposed article introduction
Imbalanced data refers to those types of datasets where the target class has an uneven distribution of observations, i.e one class label has a very high number of observations and the other has a very low number of observations. Classes that make up a large proportion of the data set are called majority classes. Those that make up a smaller proportion are minority classes. Training a model on an imbalanced dataset requires making certain adjustments otherwise the model will not perform as per your expectations.

We have various techniques that are used to handle the imbalanced dataset such undersampling technique, oversampling technique, and the SMOTE technique.

1. Undersampling technique.
In the undersampling technique, the “majority” class which is the active class will be reduced. This makes it have the same proportion as that of the “minority” class, which is the inactive class.
2. Oversampling technique.
In the oversampling technique, the “minority” class is increased so that it’s equal to the “majority” class. 
3. SMOTE technique
SMOTE is basically used to create synthetic class samples of minority classes to balance the distribution then the undersampling technique is used for cleaning irrelevant points in the boundary of the two classes to increase the separation between the two classes.

In this tutorial, we will implement each method practically when building the deep neural network. We will implement a deep learning model and add all the required layers.

#### Key takeaways 
1. Balanced vs. imbalanced datasets.
2. Impacts of building a model using an imbalanced dataset?
3. Dataset visualization.
4. One hot encoding for categorical columns
5. Build a deep neural network with Tensorflow and Keras
6. Method 1: Implementing Undersampling
7. Method2: Implementing Oversampling
8. Method3: Implementing SMOTE

#### Article quality
This tutorial is unique because we will implement each of these methods when building the deep neural network.  We will build a deep learning model for credit fraud detection. We will use metric scores such as use precision, recall, f1-score to measure the performance of our model after balancing the classes. We will compare the results with the original model that uses the imbalanced dataset to know if the model has improved.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by willyngashu**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #7272

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
Plagiarism check done ✅

**Comment by Bayler** [comment]
@willyngashu Consider shortening this URL `/implementing-undersampling-oversampling-and-smote-techniques-to-handle-imbalanced-data-in-deep-neural-networks/`. It is too long.

**Comment by willyngashu** [comment]
I have updated my article @Bayler

File: content/articles/implementing-techniques-to-handle-imbalanced-data-in-deep-neural-networks/index.md [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+```python
+dnn_pred = dnn_model.predict(X_test)
+```
+The prediction outputs are probabilities between 0 and 1 (which can be decimals). We have to round off the results to the nearest whole number. 
+
+We use the `round` method from NumPy.
+
+```python
+import numpy as np
+dnn_pred = np.round(y_preds)
```
Comment (Bayler) [inline content/articles/implementing-techniques-to-handle-imbalanced-data-in-deep-neural-networks/index.md:232]: Please confirm this code. It should be `np.round(dnn_pred)`

File: content/articles/implementing-techniques-to-handle-imbalanced-data-in-deep-neural-networks/index.md [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+- The precision has increased from 0.64 to 0.73. 
+- The recall has improved from 0.40 to 0.76. 
+- The f-score has improved from 0.49 to 0.75.
+
+These scores are also almost uniform for both classes. It ensures fair treatment for both dataset classes during the classification process.
+
+Let's move to oversampling technique.
+
+### Implementing the oversampling technique
+The undersampling technique will increase the data samples in the minority class(1) to have the same number as the majority class. It will duplicate the data samples in the minority class. We will also use the Pandas `sample` method to perform oversampling.
```
Comment (Bayler) [inline content/articles/implementing-techniques-to-handle-imbalanced-data-in-deep-neural-networks/index.md:336]: The oversampling.....

**Comment by Bayler** [comment]
@willyngashu This is a nice read. Please work on the minor revisions before we can move on for final review.

**Comment by willyngashu** [comment]
I have worked on the minor revisions @Bayler
